File: packages/design-system/src/dashboard/layout.ts
import type { Rect } from "./types.js";

const DEFAULT_RIGHT_PANE_WIDTH = 25;
const DEFAULT_FOOTER_HEIGHT = 1;
const DEFAULT_BORDER_WIDTH = 1;
const MIN_LEFT_PANE_WIDTH = 20;

export type LayoutOptions = {
  totalWidth: number;
  totalHeight: number;
  rightPaneWidth?: number;
  footerHeight?: number;
  borderWidth?: number;
};

export type DashboardLayout = {
  outerBorder: Rect;
  leftPane: Rect;
  rightPane: Rect;
  divider: { x: number; top: number; bottom: number };
  footer: Rect;
  footerDivider: { y: number; left: number; right: number };
};

export function computeDashboardLayout(opts: LayoutOptions): DashboardLayout {
  const totalWidth = normalizeSize(opts.totalWidth);
  const totalHeight = normalizeSize(opts.totalHeight);
  const borderWidth = normalizeSize(opts.borderWidth ?? DEFAULT_BORDER_WIDTH);
  const footerHeight = normalizeSize(opts.footerHeight ?? DEFAULT_FOOTER_HEIGHT);
  const requestedRightPaneWidth = normalizeSize(opts.rightPaneWidth ?? DEFAULT_RIGHT_PANE_WIDTH);
  const maxX = Math.max(0, totalWidth - 1);
  const maxY = Math.max(0, totalHeight - 1);
  const outerBorder: Rect = { x: 0, y: 0, width: totalWidth, height: totalHeight };
  const innerWidth = Math.max(0, totalWidth - (borderWidth * 2));
  const innerHeight = Math.max(0, totalHeight - (borderWidth * 2));
  const innerX = clampCoordinate(borderWidth, maxX);
  const innerY = clampCoordinate(borderWidth, maxY);
  const dividerWidth = innerWidth > 0 ? 1 : 0;
  const availablePaneWidth = Math.max(0, innerWidth - dividerWidth);
  const leftPaneWidth = computeLeftPaneWidth(availablePaneWidth, requestedRightPaneWidth);
  const rightPaneWidth = Math.max(0, availablePaneWidth - leftPaneWidth);
  const actualFooterHeight = Math.min(footerHeight, innerHeight);
  const footerDividerHeight = innerHeight > actualFooterHeight ? 1 : 0;
  const contentHeight = Math.max(0, innerHeight - actualFooterHeight - footerDividerHeight);
  const leftPane: Rect = {
    x: innerX,
    y: innerY,
    width: leftPaneWidth,
    height: contentHeight
  };
  const dividerX = clampCoordinate(leftPane.x + leftPane.width, maxX);
  const rightPane: Rect = {
    x: clampCoordinate(dividerX + dividerWidth, maxX),
    y: innerY,
    width: rightPaneWidth,
    height: contentHeight
  };
  const dividerTop = innerY;
  const dividerBottom = clampCoordinate(dividerTop + Math.max(contentHeight - 1, 0), maxY);
  const footerDividerY = clampCoordinate(innerY + contentHeight, maxY);
  const footerY = clampCoordinate(footerDividerY + footerDividerHeight, maxY);
  const footerDividerLeft = innerX;
  const footerDividerRight = clampCoordinate(Math.max(innerX, totalWidth - borderWidth - 1), maxX);

  return {
    outerBorder,
    leftPane,
    rightPane,
    divider: {
      x: dividerX,
      top: dividerTop,
      bottom: dividerBottom
    },
    footer: {
      x: innerX,
      y: footerY,
      width: innerWidth,
      height: actualFooterHeight
    },
    footerDivider: {
      y: footerDividerY,
      left: footerDividerLeft,
      right: footerDividerRight
    }
  };
}

function computeLeftPaneWidth(availablePaneWidth: number, requestedRightPaneWidth: number): number {
  if (availablePaneWidth <= 0) {
    return 0;
  }

  const maxRightPaneWidth = Math.max(0, availablePaneWidth - MIN_LEFT_PANE_WIDTH);
  const rightPaneWidth = Math.min(requestedRightPaneWidth, maxRightPaneWidth);

  return Math.max(0, availablePaneWidth - rightPaneWidth);
}

function normalizeSize(value: number): number {
  return Math.max(0, Math.floor(value));
}

function clampCoordinate(value: number, max: number): number {
  return Math.max(0, Math.min(value, max));
}

File: packages/toolcraft-design-rust/src/layout.rs
//! Portable dashboard geometry, including compact terminal behavior.
use mcp_protocol_rust::json::Value;
fn maximum(a: f64, b: f64) -> f64 {
    if a.is_nan() || b.is_nan() {
        f64::NAN
    } else {
        a.max(b)
    }
}
fn minimum(a: f64, b: f64) -> f64 {
    if a.is_nan() || b.is_nan() {
        f64::NAN
    } else {
        a.min(b)
    }
}
fn coordinate(value: f64, max: f64) -> f64 {
    maximum(0.0, minimum(value, max))
}
fn object(fields: &[(&str, f64)]) -> Value {
    Value::Object(
        fields
            .iter()
            .map(|(key, value)| (key.encode_utf16().collect(), Value::Number(*value)))
            .collect(),
    )
}
fn rect(x: f64, y: f64, width: f64, height: f64) -> Value {
    object(&[("x", x), ("y", y), ("width", width), ("height", height)])
}
/// Values are already host-coerced; Rust applies the common size/layout policy.
pub fn geometry(values: [f64; 5]) -> [f64; 25] {
    let [total_width, total_height, border, footer, requested_right] =
        values.map(|value| maximum(0.0, value.floor()));
    let max_x = maximum(0.0, total_width - 1.0);
    let max_y = maximum(0.0, total_height - 1.0);
    let inner_width = maximum(0.0, total_width - border * 2.0);
    let inner_height = maximum(0.0, total_height - border * 2.0);
    let inner_x = coordinate(border, max_x);
    let inner_y = coordinate(border, max_y);
    let divider_width = if inner_width > 0.0 { 1.0 } else { 0.0 };
    let available = maximum(0.0, inner_width - divider_width);
    let left_width = if available <= 0.0 {
        0.0
    } else {
        maximum(
            0.0,
            available - minimum(requested_right, maximum(0.0, available - 20.0)),
        )
    };
    let right_width = maximum(0.0, available - left_width);
    let footer_height = minimum(footer, inner_height);
    let footer_divider_height = if inner_height > footer_height {
        1.0
    } else {
        0.0
    };
    let content_height = maximum(0.0, inner_height - footer_height - footer_divider_height);
    let divider_x = coordinate(inner_x + left_width, max_x);
    let divider_bottom = coordinate(inner_y + maximum(content_height - 1.0, 0.0), max_y);
    let footer_divider_y = coordinate(inner_y + content_height, max_y);
    let footer_y = coordinate(footer_divider_y + footer_divider_height, max_y);
    let footer_right = coordinate(maximum(inner_x, total_width - border - 1.0), max_x);
    let compact = inner_width > 0.0 && requested_right > 0.0 && available < 40.0 + requested_right;
    let summary_height = minimum(2.0, content_height);
    [
        f64::from(compact),
        total_width,
        total_height,
        inner_x,
        if compact {
            inner_y + summary_height
        } else {
            inner_y
        },
        if compact { inner_width } else { left_width },
        if compact {
            maximum(0.0, content_height - summary_height)
        } else {
            content_height
        },
        if compact {
            max_x
        } else {
            coordinate(divider_x + divider_width, max_x)
        },
        inner_y,
        if compact { 0.0 } else { right_width },
        if compact { 0.0 } else { content_height },
        if compact { max_x } else { divider_x },
        inner_y,
        if compact { inner_y } else { divider_bottom },
        inner_x,
        footer_y,
        inner_width,
        footer_height,
        footer_divider_y,
        inner_x,
        footer_right,
        inner_x,
        inner_y,
        inner_width,
        summary_height,
    ]
}
/// Structured geometry for standalone Rust callers.
pub fn compute(values: [f64; 5]) -> Value {
    let data = geometry(values);
    let mut fields = vec![
        ("outerBorder", rect(0.0, 0.0, data[1], data[2])),
        ("leftPane", rect(data[3], data[4], data[5], data[6])),
        ("rightPane", rect(data[7], data[8], data[9], data[10])),
        (
            "divider",
            object(&[("x", data[11]), ("top", data[12]), ("bottom", data[13])]),
        ),
        ("footer", rect(data[14], data[15], data[16], data[17])),
        (
            "footerDivider",
            object(&[("y", data[18]), ("left", data[19]), ("right", data[20])]),
        ),
    ];
    if data[0] != 0.0 {
        fields.push(("summary", rect(data[21], data[22], data[23], data[24])));
    }
    Value::Object(
        fields
            .into_iter()
            .map(|(key, value)| (key.encode_utf16().collect(), value))
            .collect(),
    )
}
